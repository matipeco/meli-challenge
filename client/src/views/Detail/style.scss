.detail{
  &__container{
    background-color: #ffffff;
    padding: var(--spacing-m);
    display: flex;
    flex-wrap: wrap;
    color: var(--grey-400);
  }
  &__contain{
    max-width: 28rem;
  }
  &__image{
    width: 55rem;
  }
  &__condition{
    font-size: var(--font-200);
    margin-bottom: var(--spacing-s);
  }
  &__title{
    color: var(--grey-400);
    font-weight: 900;
    font-size: var(--font-500);
    margin-bottom: var(--spacing-m);
  }
  &__price{
    font-size: var(--font-700);
    margin-bottom: var(--spacing-m);
  }
  &__decimals{
    font-size: var(--font-500);
    margin-left: 0.5rem;
  }
  &__button{
    width: 100%;
    padding: 1rem 0;
    border: none;
    background-color: var(--blue);
    color:#ffffff;
    border-radius: 3px;
  }
  &__container-div{
    width: 100%;
  }
  &__description-title{
    display: block;
    color: var(--grey-400);
    font-size: var(--font-600);
    margin: 6rem 0 var(--spacing-m);
  }
  &__description-text{
    color: var(--grey-200);
    font-size: var(--font-300);
    &+&{
      margin-top: var(--spacing-s);
    }
  }
}