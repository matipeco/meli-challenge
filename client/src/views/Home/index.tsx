import { Container } from "../../components/Container"
import './style.scss'
import landing from '../../assets/meli-background.png'

export const Home =()=>{
  return(
    <main className="landing__container">
      <Container>
        <img className="landing__background" src={landing} alt="mercadolibre landing" width="500" />
      </Container>
    </main>
  )
}