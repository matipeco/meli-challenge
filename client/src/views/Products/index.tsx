import { FunctionComponent, useEffect, useState } from "react"
import { Container } from "../../components/Container"
import axios from 'axios';
import { SearchResponse } from "../../types";
import { useLocation } from "react-router-dom";
import { Card } from "../../components/Card";
import './style.scss'

export const Products:FunctionComponent = () => { 
  const { search } = useLocation();
  const params = new URLSearchParams(search);
  const q = params.get('q');

  const [info, setInfo] = useState<SearchResponse>();

  const getProducts = async() =>{
    try {
      const { data } = await axios.get<SearchResponse>(`http://localhost:3001/api/items?q=${q}`);
      setInfo(data);
      console.log(data)

    } catch (error) {
      console.log(error);
    }
  }

  useEffect( ()=>{
    getProducts()
  },[q])

  {console.log(info?.categories)}
  return(
   <main className="container-cards">
    <Container>
        {info?.categories.map((cat)=>{
          return(
            <span className="categories">{cat} - </span>
          )
        })}
      {info?.items.map((prod)=>{
        return(
          <Card prod={prod}/>
        )
      })}
    </Container>
   </main> 
  )
}