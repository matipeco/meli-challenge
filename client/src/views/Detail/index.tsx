import { useParams } from "react-router-dom";
import { Container } from "../../components/Container"
import { useEffect, useState } from "react";
import axios from "axios";
import { DetailProduct, DetailResponse } from "../../types";
import { Breadcrumb } from "../../components/Breadcrumb";
import { Spinner } from "../../components/Spinner";
import './style.scss';
import { Status } from "../Products";
import { ProductNotFound } from "../../components/ProductNotFound";
import { descriptionArray, getDecimals, priceFormat } from "../../utils";

export const Detail = () => {
  const { id } = useParams();

  const [product, setProduct] = useState<DetailProduct>();
  const [categories, setCategories] = useState<string[]>([]);
  const [status, setStatus] = useState<Status>("initial");
  
  const getDetail = async() => {
    try {
      setStatus("loading")
      const { data } = await axios.get<DetailResponse>(`http://localhost:3001/api/items/${id}`);
      setStatus("success")
      setProduct(data.item)
      setCategories(data.categories)
    } catch (error) {
      setStatus("error")
    }
  }
  
  useEffect(()=>{
    getDetail();
  },[])
  
  return(
    <main className="main__detail">
      <Container>
      {status === "loading" && <Spinner/>}
      {status === "error" && <ProductNotFound/>}
      {status === "success" && product && (
      <>
          <Breadcrumb categories={categories}/>
          <article className="detail__container">
            <img className="detail__image" src={product.picture} alt={product.title} />
            
            <div className="detail__contain">
              <p className="detail__condition">{product.condition} - {product.sold_quantity} {product.sold_quantity === 1 ? "vendido" : "vendidos"} </p>
              <h2 className="detail__title">
                {product.title}
              </h2>
              <p className="detail__price">{priceFormat(product)}
                <sup className="detail__decimals">{getDecimals(product.price.decimals)}</sup> 
              </p>
              <button className="detail__button">Comprar</button>
            </div>
            <div className="detail__container-div">
              {descriptionArray.length !== 0 && 
                <h3 className="detail__description-title">Descripci√≥n del producto</h3>
                }
              {descriptionArray(product).map((paragraph)=>{
                return(
                  <p className="detail__description-text" key={paragraph}>{paragraph}</p>
                )
              })}
            </div>
          </article>
      </>
      )}
      </Container>
    </main>
  )
}
