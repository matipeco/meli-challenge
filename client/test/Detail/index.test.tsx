import {waitForElementToBeRemoved, render, screen,} from '@testing-library/react';
import { Detail } from '../../src/views/Detail'
import { BrowserRouter } from 'react-router-dom';
import axios from 'axios'

describe("Componente Detail", ()=>{
  const axiosSpy = vi.spyOn(axios, "get").mockResolvedValue({
    data: {
      author: {
        name: "Matias",
        lastname:"Pecorale"
      },
      categories:["Celulares", "Iphone"],
      item: {
        id: "MLA1408529234",
        title: "Apple iPhone 11 (128 Gb) - Blanco",
        price: {
          currency: "ARS",
          amount: 340000,
          decimals: 0
        },
        picture: "http://http2.mlstatic.com/D_796892-MLA46114829828_052021-I.jpg",
        condition: "new",
        free_shipping: true,
        sold_quantity: 454545,
        description: "asdasda"
      }
    }
  })

  test("Se renderiza correctamente", async()=>{
    const {container} = render(
      <Detail/>, {wrapper:BrowserRouter}
    )

    const button = await screen.findByRole("button");

    expect(container).toMatchSnapshot();
  })

  test("Renderiza ProductNotFound si no se encuentra el producto", async()=>{
    axiosSpy.mockRejectedValueOnce({
      response:{
        status:404
      }
    })
    const {container} = render(
      <Detail/>, {wrapper:BrowserRouter}
    )
    
    await waitForElementToBeRemoved(screen.getByAltText("cargando"))

    expect(container).toMatchSnapshot();
  })
})