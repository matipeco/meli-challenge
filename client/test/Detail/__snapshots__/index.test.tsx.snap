// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Componente Detail > Renderiza ProductNotFound si no se encuentra el producto 1`] = `
<div>
  <main
    class="main__detail"
  >
    <div
      class="container"
    >
      <div
        class="product-not__div"
      >
        <article
          class="product-not__article"
        >
          <img
            alt=""
            height="20"
            src="/src/assets/productnot.svg"
            width="20"
          />
          <div>
            <p>
              No hay publicaciones que coincidan con tu búsqueda
            </p>
            <ul>
              <li>
                Revisá la ortografía de la palabra.
              </li>
              <li>
                Utilizá palabras más genéricas o menos palabras.
              </li>
            </ul>
          </div>
        </article>
      </div>
    </div>
  </main>
</div>
`;

exports[`Componente Detail > Se renderiza correctamente 1`] = `
<div>
  <main
    class="main__detail"
  >
    <div
      class="container"
    >
      <ol
        class="breadcrumb"
      >
        <li>
          <a
            href="#"
          >
            Celulares
          </a>
        </li>
        <li>
          <a
            href="#"
          >
            Iphone
          </a>
        </li>
      </ol>
      <article
        class="detail__container"
      >
        <img
          alt="Apple iPhone 11 (128 Gb) - Blanco"
          class="detail__image"
          src="http://http2.mlstatic.com/D_796892-MLA46114829828_052021-I.jpg"
        />
        <div
          class="detail__contain"
        >
          <p
            class="detail__condition"
          >
            new
             - 
            454545
             
            vendidos
             
          </p>
          <h2
            class="detail__title"
          >
            Apple iPhone 11 (128 Gb) - Blanco
          </h2>
          <p
            class="detail__price"
          >
            $ 340.000
            <sup
              class="detail__decimals"
            >
              00
            </sup>
          </p>
          <button
            class="detail__button"
          >
            Comprar
          </button>
        </div>
        <div
          class="detail__container-div"
        >
          <h3
            class="detail__description-title"
          >
            Descripción del producto
          </h3>
          <p
            class="detail__description-text"
          >
            asdasda
          </p>
        </div>
      </article>
    </div>
  </main>
</div>
`;
